internal/contributebot/main.go

```diff
@@ -106,94 +107,102 @@ func (w *worker) receive(ctx context.Context) error {
 }
 
 func (w *worker) receiveIssueEvent(ctx context.Context, e *github.IssuesEvent) error {
-	// An issueRule is a rule about what to do for a given event.
-	// Rules are executed in random order to ensure there are no dependencies.
-	type issueRule struct {
-		// name is the rule's name for error messages.
-		name string
-
-		// condition reports true if the event is "interesting" to this rule,
-		// meaning that the worker should fetch the latest information about
-		// the issue and call the run function.
-		condition func(*github.IssuesEvent) bool
-
-		// run executes the rule. The issue data is retrieved before any
-		// rules are executed, so may be ever-so-slightly stale.
-		run func(context.Context, *github.Client, issueRuleData) error
-	}
 
-	rules := []issueRule{
-		// Remove "in progress" label from closed issues.
-		{
-			name: "remove in progress label from closed",
-			condition: func(e *github.IssuesEvent) bool {
-				return e.GetAction() == "closed" && hasLabel(e.GetIssue(), inProgressLabel)
-			},
-			run: removeInProgressLabel,
-		},
+	data := issueRuleData{
+		action:  e.GetAction(),
+		owner:   e.GetRepo().GetOwner().GetLogin(),
+		repo:    e.GetRepo().GetName(),
+		issue:   e.GetIssue(),
+		changes: e.GetChanges(),
 	}
 
 	// Check conditions to see if issue data is needed.
 	// No need to consume API quota if events aren't relevant.
-	runs := make([]bool, len(rules))
-	runCount := 0
-	for i := range rules {
-		if rules[i].condition(e) {
-			runs[i] = true
-			runCount++
+	var toRun []issueRule
+	for _, rule := range allIssueRules {
+		if rule.Condition(data) {
+			toRun = append(toRun, rule)
 		}
 	}
-	if runCount == 0 {
+	if len(toRun) == 0 {
+		log.Printf("No issue rules matched %v", data)
 		return nil
 	}
 
-	// Retrieve the current issue state.
 	client := w.ghClient(e.GetInstallation().GetID())
-	owner := e.GetRepo().GetOwner().GetLogin()
-	repoName := e.GetRepo().GetName()
-	num := e.GetIssue().GetNumber()
-	iss, _, err := client.Issues.Get(ctx, owner, repoName, num)
-	if err != nil {
-		return err
-	}
 
 	// Execute relevant rules.
 	ok := true
-	data := issueRuleData{
-		issue: iss,
-		owner: owner,
-		repo:  repoName,
-	}
-	for i := range runs {
-		if !runs[i] {
+	for _, r := range toRun {
+		// Retrieve the current issue state after every rule,
+		// so that subsequent rules have the latest issue data.
```

Comment:
1. The current code states that the rules are run in random order, but it looks to me like they are always run in the same order (based on the "rules" slice). There aren't any maps to make things random.
1. I'm not sure I agree that order-dependent is harder to reason about.
- If all of the rules are truly independent, then it doesn't matter what order they run in.
- If the rules are *not* truly independent, then the order does matter, and it seems like you'd want to be able to control it.

As an example, for other issues I'm adding
- A check that will close a PR if it's from a branch of the main repo.
- Checks to do a bunch of other stuff on the PR.
If the first check fires and closes the PR, we probably don't want to do all of the other checks.

Reply:
Rephrasing:

I think you are asserting that it would be easier to reason about if we made it a requirement that all rules are independent (i.e., nothing one rule does can affect another rule's condition or action). If that were true, it wouldn't matter what order the rules run in. However, I'm not sure it's a desirable requirement (e.g., see the example in my previous comment), and it would be difficult to enforce anyway since rules can do arbitrary things in condition/action. If rules **can** affect one another, then I think it's easier to reason about them if they are in a stable order.